### ISSUE TITLE
`kpt live apply` creates inventory entries when access is forbidden

### ISSUE THREAD

**Issue by fsommar** [GoogleContainerTools/kpt#1192]
**TL;DR:** `kpt live apply` will add resources it doesn't have access to to its inventory, causing all further interactions with `kpt live` to fail. I'm expecting resources that cannot be created to not be added to the inventory.

In the spirit of including a fully contained background of what's happening, I'll include a section on how to setup Kind to mirror the setup that we're using for our multi-tenant clusters with namespaced-based access control. If you have something similar set up already, feel free to skip the following section.

## Setup

I'm assuming that kind is installed and that KUBECONFIG is set to something different to avoid clobbering the user KUBECONFIG - it should only be additive, though.

```bash
> kind create cluster --name=tmp
> cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: Namespace
metadata:
  name: test
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: user
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: admin
  namespace: test
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cluster-admin
subjects:
- kind: User
  name: system:serviceaccount:default:user
  apiGroup: rbac.authorization.k8s.io
EOF
> kubectl config set-credentials user --token="$(kubectl get secrets -ojsonpath='{.data.token}' \
  "$(kubectl get sa user -ojsonpath='{.secrets[0].name}')" \
  | base64 -d)"
> kubectl config set-context kind-tmp:user --cluster=kind-tmp --user=user
> kubectl config use-context kind-tmp:user
```

## Actual

At this point, the following outputs are expected:

```sh
> kubectl auth can-i create configmap -n test
yes
> kubectl auth can-i create configmap
no
```

Let's add a configmap, run `kpt live init`, and apply it.

```sh
> cd "$(mktemp -d)"
> cat <<EOF > cm.yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: cm
  namespace: test
data: {}
EOF
> kpt live init .
namespace: test is used for inventory object
Initialized: /var/folders/tt/_ps2bxp912d840z7hvjp4t100000gn/T/tmp.De6gBVnq/inventory-template.yaml
> kpt live apply .
configmap/cm created
1 resource(s) applied. 1 created, 0 unchanged, 0 configured
0 resource(s) pruned, 0 skipped
> kubectl -n test get cm -lcli-utils.sigs.k8s.io/inventory-id -ojsonpath='{.items[*].data}'
{"test_cm__ConfigMap":""}
```

The inventory now contains one item in the `test` namespace, as expected. Now, let's add another configmap, this time accidentally forgetting to set the namespace.

```sh
> cat <<EOF > empty.yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: empty
data: {}
EOF
> kpt live apply .
Error from server (Forbidden): error when retrieving current configuration of:
Resource: "/v1, Resource=configmaps", GroupVersionKind: "/v1, Kind=ConfigMap"
Name: "empty", Namespace: "default"
Object: &{map["apiVersion":"v1" "data":map[] "kind":"ConfigMap" "metadata":map["annotations":map["kubectl.kubernetes.io/last-applied-configuration":""] "name":"empty" "namespace":"default"]]}
from server for: "empty.yaml": configmaps "empty" is forbidden: User "system:serviceaccount:default:user" cannot get resource "configmaps" in API group "" in the namespace "default"
> kubectl -n test get cm -lcli-utils.sigs.k8s.io/inventory-id -ojsonpath='{.items[*].data}' 
{"default_empty__ConfigMap":"","test_cm__ConfigMap":""}
```

Oops. That didn't work. Let's delete `empty.yaml` and try again.

```sh
> rm empty.yaml
> kpt live apply .
configmap/cm configured
error: polling for status failed: configmaps is forbidden: User "system:serviceaccount:default:user" cannot list resource "configmaps" in API group "" in the namespace "default"
```

It seems that our inventory is now tainted; any actions using `kpt live` will fail since it's unable to poll its inventory items.

## Expected

My expectation in this scenario is that `kpt live apply` should not add resources to its inventory that it wasn't able to create. The first `kpt live apply` with the erroneous resource should fail, and the inventory should look the same as before the call. Thus deleting the resource and re-running `kpt live apply` should work.

Our current work-around is to manually remove the offending resource from the inventory on behalf of our users, but there's only so many things that we want to do manually on behalf of our users 😄

**Comment by mortent** [comment]
I have looked into this and it raises many interesting questions. It did expose a weakness in the polling functionality where an unknown error during the sync phase of the caching reader would cause it to bail out. I have a [PR](https://github.com/kubernetes-sigs/cli-utils/pull/279) to fix this as it is pretty clear it isn't the correct behavior.

But the issue remains even with this change. I think there are several interesting questions around how we should handle this:

* Can we detect whether a resource can be created before we add it to the inventory?
  * We could do a server-side dry-run, but it could be a bit expensive. Even then, are we guaranteed to catch all possible errors?
  * We can end up in this situation anyway since RBAC permissions might change after a resource was applied.
* Can we clean it up during apply, so after we detect one or more resources couldn't be applied, we remove them from the inventory?
  * Have to be careful to make sure we don't remove entries for resources that already exists. This is true even if the RBAC permissions prevent us from fetching it.
* Can we continue with pruning after one or more resources couldn't be applied? If we can't look up the resource, we don't know the UID.
* Should we remove entries from the inventory if we can't look up the resource? It means we don't know whether it exists or not. Removing the entry from the inventory means orphaning any existing resources.
* Is there ever a safe way to prune a resource if we can't look it up? This would require some strange RBAC permissions, but we need to look at the resource in order to support lifecycle-directives.

@seans3 @Liujingfang1

**Comment by mortent** [comment]
@fsommar Thanks for the detailed description! This is very helpful.

**Comment by seans3** [comment]
Hi @fsommar . Thanks for reporting this interesting case. And especially thanks for the very detailed steps to reproduce this problem; that really helps.

There is a bug here. But it is not the one you report. The problem is not that we've stored an item in the inventory that we could not successfully apply. This is a relatively common occurrence that the algorithm is tolerant to. The problem is that the current implementation erroneously stops when encountering the polling error. This cessation blocks the final step of the algorithm which stores the currently applied items in the inventory. The algorithm must always "continue on error". Once the algorithm correctly stores the applied inventory (even when it encounters errors), the final step in your example will correctly remove the extra item from the inventory.

**Comment by fsommar** [comment]
Thank you for your responses! The 4-step description helped cement what's happening here for me.

> Should we remove entries from the inventory if we can't look up the resource? It means we don't know whether it exists or not. Removing the entry from the inventory means orphaning any existing resources.

I would prefer that, but we're also resilient to orphaned resources so it wouldn't cause us any direct issues other than a source of confusion for our users. Given our setup, I don't see that scenario ever happening unless we change RBAC for resources already managed in an inventory.

> And if there are extra or spurious items in the inventory during an apply, the algorithm will simply attempt to prune/delete items which do not exist--which is not a problem.

Each apply would attempt to prune these resources, correct? As the prune would never succeed, and so the resource would not be removed from the inventory. That is, unless the above method is applied, removing resources that are inaccessible and possibly (while unlikely) orphaning them.

**Comment by seans3** [comment]
Final step to fix this: https://github.com/GoogleContainerTools/kpt/issues/1192

**Comment by seans3** [comment]
As of the latest cli-utils code `v0.22.5`, kpt produces the following output when applying both ConfigMaps:
```
$ kpt live apply .
configmap/empty unchanged
configmap/cm created
2 resource(s) applied. 1 created, 1 unchanged, 0 configured, 0 failed
0 resource(s) pruned, 0 skipped, 0 failed
$ kubectl get cm -n test
NAME                 DATA   AGE
cm                   0      14s
inventory-86135376   1      14s
```
The inventory does not store the ConfigMap name `empty` which has RBAC errors; it only stores the correct ConfigMap named `cm`. The output from apply `configmap/empty unchanged` and that `0 failed`  is wrong and must be fixed. However, if we dive deeper into what is going on, we see that the apply failed as it should and the erroneous ConfigMap was not ultimately stored in the inventory:
```
$ kpt live apply . -v=5
I0201 23:23:42.825479  624662 applycmd.go:71] wrapper applyRunner run...
I0201 23:23:42.832042  624662 inventory-client.go:247] prune inventory object fetch: configmaps/test/cli-utils.sigs.k8s.io/inventory-id=7bbb7567-fe89-4821-9efb-c4c2e22106d5
I0201 23:23:42.840673  624662 applier.go:180] apply run for 2 objects
I0201 23:23:42.840697  624662 applier.go:93] applier preparing 2 objects
I0201 23:23:42.840707  624662 applier.go:108] applier merging 2 objects into inventory
I0201 23:23:42.840738  624662 inventory-client.go:247] prune inventory object fetch: configmaps/test/cli-utils.sigs.k8s.io/inventory-id=7bbb7567-fe89-4821-9efb-c4c2e22106d5
I0201 23:23:42.843141  624662 inventory-client.go:118] creating initial inventory object with 2 objects
I0201 23:23:42.843531  624662 inventory-client.go:392] creating inventory object: test/inventory-86135376
I0201 23:23:42.846655  624662 applier.go:117] after inventory merge; 0 objects to prune
I0201 23:23:42.846674  624662 applier.go:203] applier building task queue...
I0201 23:23:42.846694  624662 applier.go:238] applier building TaskStatusRunner...
I0201 23:23:42.846705  624662 applier.go:240] applier running TaskStatusRunner...
I0201 23:23:42.846731  624662 apply_task.go:77] apply task starting; attempting to apply 2 objects
I0201 23:23:42.847254  624662 apply_task.go:125] attempting to apply 2 remaining objects
E0201 23:23:42.848155  624662 apply_task.go:147] error retrieving default/empty from cluster--continue
configmap/empty unchanged
I0201 23:23:42.850322  624662 apply_task.go:173] applying test/cm...
configmap/cm created
2 resource(s) applied. 1 created, 1 unchanged, 0 configured, 0 failed
I0201 23:23:42.853516  624662 inventory-client.go:247] prune inventory object fetch: configmaps/test/cli-utils.sigs.k8s.io/inventory-id=7bbb7567-fe89-4821-9efb-c4c2e22106d5
I0201 23:23:42.857067  624662 prune.go:118] prune: 2 objects attempted to apply
I0201 23:23:42.857085  624662 prune.go:119] prune: 1 objects successfully applied
I0201 23:23:42.857092  624662 prune.go:120] prune: 2 union objects stored in cluster inventory
I0201 23:23:42.857106  624662 prune.go:122] prune: 0 objects to prune (clusterInv - localIds)
E0201 23:23:42.857568  624662 prune.go:204] error retrieving object for inventory determination: configmaps "empty" is forbidden: User "system:serviceaccount:default:user" cannot get resource "configmaps" in API group "" in the namespace "default"
I0201 23:23:42.858586  624662 prune.go:210] adding final inventory object test/cm
I0201 23:23:42.858599  624662 prune.go:217] final inventory 1 successfully applied objects
I0201 23:23:42.858607  624662 prune.go:219] final inventory 1 objects after appending prune failures
I0201 23:23:42.858623  624662 inventory-client.go:247] prune inventory object fetch: configmaps/test/cli-utils.sigs.k8s.io/inventory-id=7bbb7567-fe89-4821-9efb-c4c2e22106d5
I0201 23:23:42.860843  624662 inventory-client.go:247] prune inventory object fetch: configmaps/test/cli-utils.sigs.k8s.io/inventory-id=7bbb7567-fe89-4821-9efb-c4c2e22106d5
I0201 23:23:42.863929  624662 inventory-client.go:179] replace cluster inventory: test/inventory-86135376
I0201 23:23:42.863940  624662 inventory-client.go:180] replace cluster inventory 1 objects
I0201 23:23:42.864100  624662 inventory-client.go:359] replacing inventory object: test/inventory-86135376
0 resource(s) pruned, 0 skipped, 0 failed
```

The problem where unsuccessfully applied objects ended up in the inventory is fixed. The apply error output, however, needs to be fixed.

**Comment by seans3** [comment]
This is now fixed. Output for this case is:
```
$ kpt live apply .
configmap/empty failed
configmap/cm created
2 resource(s) applied. 1 created, 0 unchanged, 0 configured, 1 failed
0 resource(s) pruned, 0 skipped, 0 failed
$ kubectl get cm -n test
NAME                 DATA   AGE
cm                   0      2m19s
inventory-01050827   1      2m19s
$ kubectl get cm/empty
Error from server (Forbidden): configmaps "empty" is forbidden: User "system:serviceaccount:default:user" cannot get resource "configmaps" in API group "" in the namespace "default"
```
Only one successful apply is in the inventory, and the apply correctly reported the `empty` config as `failed`.

We will close this when the following PR is merged: https://github.com/kubernetes-sigs/cli-utils/pull/322
and the newly built `cli-utils` version is integrated into `kpt`. This use case will be integrated into the end-to-end test to ensure no regressions.

**Comment by seans3** [comment]
The fix will be in `cli-utils` version `v0.23.0`

**Comment by seans3** [comment]
Fixed in `kpt` built at master. I will update this issue when we do the next release of `kpt`. Closing.
